File: parsers/hlb_csv.py
from __future__ import annotations

import csv
import re
from datetime import datetime

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
from categorize import Transaction


_REF_ID = re.compile(r'\d{8}HLBBMYKL\w+$')   # trailing reference ID in DuitNow / transfer refs
_SPACES = re.compile(r'\s{2,}')               # multiple spaces → single space


def _clean_ref(desc_type: str, reference: str) -> str:
    t = desc_type.upper()
    r = reference.strip()

    if 'POS PURCHASE' in t or 'MYDEBIT' in t:
        # "DEBITCARD 7-ELEVEN MALAYSIA SDN BH ISKANDAR PUTEMY5499"
        # strip "DEBITCARD " prefix; MyDebit refs have no prefix at all
        r = re.sub(r'^DEBITCARD\s+', '', r, flags=re.IGNORECASE)

    elif 'DUITNOW QR' in t:
        # "QR PaymentGMO(TAMANEKOBOTANI)20260712HLBBMYKL030OQR39684924"
        # strip "QR Payment" prefix then trailing ref ID
        r = re.sub(r'^QR Payment', '', r, flags=re.IGNORECASE)
        r = _REF_ID.sub('', r)

    elif 'HL CONNECT' in t or 'INSTANT TRANSFER' in t or 'DUITNOW' in t:
        # "Fund transferEU YU HAN20260710HLBBMYKL010ORM38764234"
        # strip "Fund transfer" prefix then trailing ref ID
        r = re.sub(r'^Fund transfer', '', r, flags=re.IGNORECASE)
        r = _REF_ID.sub('', r)

    return _SPACES.sub(' ', r).strip()


def parse(filepath: str) -> list[Transaction]:
    transactions: list[Transaction] = []

    with open(filepath, encoding='utf-8-sig', newline='') as f:
        reader = csv.DictReader(f)
        for i, row in enumerate(reader, start=2):
            wd_raw = (row.get('Withdrawal (MYR)') or '').strip().replace(',', '')
            if not wd_raw:
                continue  # deposit / credit row

            amount = float(wd_raw)
            if amount <= 0:
                continue

            date_str = (row.get('Date') or '').strip()
            try:
                date = datetime.strptime(date_str, '%d-%b-%Y').strftime('%Y-%m-%d')
            except ValueError:
                raise ValueError(f"Row {i}: expected DD-Mon-YYYY date, got {date_str!r}")

            desc_type = (row.get('Description') or '').strip()
            reference = (row.get('Reference') or '').strip()
            cleaned   = _clean_ref(desc_type, reference)
            description = f"{desc_type} / {cleaned}" if cleaned else desc_type

            transactions.append(Transaction(
                date=date,
                description=description,
                amount=amount,
                currency='MYR',
                payment_type='DuitNow QR / POS / Bank transfer',
                source_account='HLB Pay & Save',
            ))

    return transactions

File: webapp/src/parsers/hlbCsv.js
import Papa from 'papaparse';
import { makeTransaction } from '../categorize.js';

const REF_ID_RE = /\d{8}HLBBMYKL\w+$/;
const SPACES_RE = /\s{2,}/g;

const MONTHS = {
  jan: 1, feb: 2, mar: 3, apr: 4, may: 5, jun: 6,
  jul: 7, aug: 8, sep: 9, oct: 10, nov: 11, dec: 12,
};

function cleanRef(descType, reference) {
  const t = descType.toUpperCase();
  let r = reference.trim();

  if (t.includes('POS PURCHASE') || t.includes('MYDEBIT')) {
    // "DEBITCARD 7-ELEVEN MALAYSIA SDN BH ISKANDAR PUTEMY5499"
    // strip "DEBITCARD " prefix; MyDebit refs have no prefix at all
    r = r.replace(/^DEBITCARD\s+/i, '');
  } else if (t.includes('DUITNOW QR')) {
    // "QR PaymentGMO(TAMANEKOBOTANI)20260712HLBBMYKL030OQR39684924"
    // strip "QR Payment" prefix then trailing ref ID
    r = r.replace(/^QR Payment/i, '');
    r = r.replace(REF_ID_RE, '');
  } else if (t.includes('HL CONNECT') || t.includes('INSTANT TRANSFER') || t.includes('DUITNOW')) {
    // "Fund transferEU YU HAN20260710HLBBMYKL010ORM38764234"
    // strip "Fund transfer" prefix then trailing ref ID
    r = r.replace(/^Fund transfer/i, '');
    r = r.replace(REF_ID_RE, '');
  }

  return r.replace(SPACES_RE, ' ').trim();
}

function parseDdMonYyyy(dateStr) {
  const m = /^(\d{2})-([A-Za-z]{3})-(\d{4})$/.exec(dateStr);
  if (!m) return null;
  const month = MONTHS[m[2].toLowerCase()];
  if (!month) return null;
  return `${m[3]}-${String(month).padStart(2, '0')}-${m[1]}`;
}

/**
 * @param {File} file
 * @returns {Promise<import('../categorize.js').Transaction[]>}
 */
export async function parse(file) {
  const rawText = await file.text();
  const text = rawText.replace(/^﻿/, ''); // utf-8-sig BOM
  const { data } = Papa.parse(text, { header: true, skipEmptyLines: true });

  const transactions = [];
  data.forEach((row, i) => {
    const wdRaw = (row['Withdrawal (MYR)'] || '').trim().replace(/,/g, '');
    if (!wdRaw) return; // deposit / credit row

    const amount = parseFloat(wdRaw);
    if (amount <= 0) return;

    const dateStr = (row['Date'] || '').trim();
    const date = parseDdMonYyyy(dateStr);
    if (!date) {
      throw new Error(`Row ${i + 2}: expected DD-Mon-YYYY date, got ${JSON.stringify(dateStr)}`);
    }

    const descType = (row['Description'] || '').trim();
    const reference = (row['Reference'] || '').trim();
    const cleaned = cleanRef(descType, reference);
    const description = cleaned ? `${descType} / ${cleaned}` : descType;

    transactions.push(makeTransaction({
      date,
      description,
      amount,
      currency: 'MYR',
      payment_type: 'DuitNow QR / POS / Bank transfer',
      source_account: 'HLB Pay & Save',
    }));
  });

  return transactions;
}
